core/src/test/scala/unit/kafka/server/BaseRequestTest.scala

```diff
@@ -84,104 +83,46 @@ abstract class BaseRequestTest extends IntegrationTestHarness {
     outgoing.flush()
   }
 
-  private def receiveResponse(socket: Socket): Array[Byte] = {
+  def receive(socket: Socket, apiKey: ApiKeys, version: Short): AbstractResponse = {
     val incoming = new DataInputStream(socket.getInputStream)
     val len = incoming.readInt()
+
     val response = new Array[Byte](len)
     incoming.readFully(response)
-    response
+
+    val responseBuffer = ByteBuffer.wrap(response)
+    ResponseHeader.parse(responseBuffer, apiKey.responseHeaderVersion(version))
+
+    val responseStruct = apiKey.parseResponse(version, responseBuffer)
+    AbstractResponse.parseResponse(apiKey, responseStruct, version)
   }
 
-  def requestAndReceive(socket: Socket, request: Array[Byte]): Array[Byte] = {
-    sendRequest(socket, request)
-    receiveResponse(socket)
+  def sendAndReceive(request: AbstractRequest, socket: Socket): AbstractResponse = {
```

Comment:
You can add a type parameter for the desired response type along with a ClassTag so that you can pattern match safely and either return the desired type or throw an exception. Then the callers no longer need to cast.

Reply:
Thanks for the suggestion. Updated.